CMake: fixes dependencies of iso9660++

Karlson2k opened the pull request:
**The PR only fixes listing of `libiso9660++` dependencies in correct order.**

By request of @wsnipex added check to match version of `libcdio` and `libiso9660++` (as they should be built from the same sources).
No new features, no changes of mandatory/optional dependency are introduced.

More details:
`libiso9660++` is built from `libcdio` sources, but it is optional depending on configuration parameters, see http://git.savannah.gnu.org/cgit/libcdio.git/tree/configure.ac?h=release-2.1.0#n74.
In overall, up to 5 libraries can be built from libcdio sources: 
* [`libcdio`](http://git.savannah.gnu.org/cgit/libcdio.git/tree/libcdio.pc.in?h=release-2.1.0), 
* [`libcdio++`](http://git.savannah.gnu.org/cgit/libcdio.git/tree/libcdio++.pc.in?h=release-2.1.0),
* [`libiso9660`](http://git.savannah.gnu.org/cgit/libcdio.git/tree/libiso9660.pc.in?h=release-2.1.0), 
* [`libiso9660++`](http://git.savannah.gnu.org/cgit/libcdio.git/tree/libiso9660++.pc.in?h=release-2.1.0),
* [`libudf`](http://git.savannah.gnu.org/cgit/libcdio.git/tree/libudf.pc.in?h=release-2.1.0).

See [libcdio sources](http://git.savannah.gnu.org/cgit/libcdio.git/tree/?h=release-2.1.0). 

Linux distributions sometimes make several packages from the same sources. See [Ubuntu](https://packages.ubuntu.com/search?keywords=libcdio&searchon=sourcenames&suite=groovy&section=all) and [Debian](https://packages.debian.org/source/buster/libcdio).

1. Kodi uses `libcdio` for direct access of optical devices, audio CD playback and guessing FS type of optical disks.
2. `libiso9660++` is **optionally** used by Kodi for reading of some .iso images and some CD-ROM and Video CD disks.
`libiso9660++` depends on `libcdio++` headers, `libiso9660` headers and binary, `libcdio` headers and binary.

`libudf` is not used by Kodi (we are using alternative `libudfread`).

Karlson2k committed:
CMake: fixes dependencies of iso9660++

iso9660++ require cdio libs. Enable iso9660++ only if all
dependencies are found.
Add all dependencies in proper order to the lists of libs and include dirs.

wsnipex reviewed (changes requested):
suffers from the same issues as the other PR

Karlson2k commented:
Hi @wsnipex.

Let's list the issues:
>* There is no check if cdio.hpp is actually available

It is checked here:
https://github.com/xbmc/xbmc/pull/19234/commits/b5e5379a5ba5714cac0dfa29cea61ac31735eb07#diff-88004735dc842191b2b75b219193c7c288422f0cfa1d567ed0dc2519f3db7695R27

Required here: https://github.com/xbmc/xbmc/pull/19234/commits/b5e5379a5ba5714cac0dfa29cea61ac31735eb07#diff-88004735dc842191b2b75b219193c7c288422f0cfa1d567ed0dc2519f3db7695R44

Included here: https://github.com/xbmc/xbmc/pull/19234/commits/b5e5379a5ba5714cac0dfa29cea61ac31735eb07#diff-88004735dc842191b2b75b219193c7c288422f0cfa1d567ed0dc2519f3db7695R49

>* iso9660++ is required for ISO playback, therefore we also need cdio++ and cdio

That's correct. To enable `iso9660++`, we need to check for `libcdio`, `libcdio++`, and `iso9660`.
They are checked here (cdio): https://github.com/xbmc/xbmc/pull/19234/commits/b5e5379a5ba5714cac0dfa29cea61ac31735eb07#diff-88004735dc842191b2b75b219193c7c288422f0cfa1d567ed0dc2519f3db7695R19
here (cdio++): https://github.com/xbmc/xbmc/pull/19234/commits/b5e5379a5ba5714cac0dfa29cea61ac31735eb07#diff-88004735dc842191b2b75b219193c7c288422f0cfa1d567ed0dc2519f3db7695R27
and here (iso9660): https://github.com/xbmc/xbmc/pull/19234/commits/b5e5379a5ba5714cac0dfa29cea61ac31735eb07#diff-88004735dc842191b2b75b219193c7c288422f0cfa1d567ed0dc2519f3db7695R34

>* don't mix and match Find modules. Everything cdio related should be handled in FindCdio.cmake

This makes sense. `libcdio++` is not required to use `libcdio`, but `libcdio++` is required for `iso9660++`. Therefore `libcdio++` is handled in `FindIso9660pp.cmak`. It can be moved to the `FindCdiopp.cmake` for clarity, but it can be overcomplication.

Karlson2k commented:
This is a fix only!
No new features, no enhancements.
Only fix for correct detection of `iso9660` during configuration phase to avoid failing at build stage.

Karlson2k committed:
CMake: added check for iso9660++ and cdio version match

This is just a precaution as it's hardly possible that libraries built
from the same sources will have different versions.

Karlson2k commented:
@DaveTBlake I believe this PR is useful for release and doesn't change any mandatory/optional state of Kodi dependencies.

@wsnipex requested to make a hard dependency of `libcdo++`, which I found pointless as `libcdio++` is not used by Kodi directly and used only by `iso9660++` which is still optional.

lrusak commented on cmake/modules/FindIso9660pp.cmake:
```diff
@@ -12,31 +12,47 @@
 find_library(ISO9660PP_LIBRARY NAMES libiso9660++ iso9660++
                                PATHS ${PC_ISO9660PP_LIBDIR})
 
+find_path(CDIOPP_INCLUDE_DIR NAMES cdio++/cdio.hpp
```
I don't really understand why you want to check for a header that we don't explicitly need?

Karlson2k replied:
It's the first `#include` in `libiso9660++` header, i.e. it's used indirectly by `libiso9660++`.

yol replied:
I really think that is something that distros have to take care of. I.e., not our problem.

yol replied:
(AFAIK we don't do this for other dependencies, as well)

Karlson2k replied:
Yep, proper dependencies in distro packages should works fine. Actually it could be removed.

Karlson2k replied:
Initially it was added because I had added library binary file as well, but later it was discovered that binary is not required.

DaveTBlake commented:
@Karlson2k thank you for looking at this. I have been watching the back and forth on this topic, and really don't know what to conclude. As I understand it what is in master works, maybe by accident, but it works and that matters!

We are hours before release and hence would want to see some team consensus if it is to go in.

Karlson2k commented:
@DaveTBlake If in doubt, skip it. If packages in distribution has correct dependencies `libiso9660++` will be used (or not used if it's not available).
The current combination of CMake files and modules provides working combination of dependencies and will not break unless we touch them.

DaveTBlake commented:
>  will not break unless we touch them

That sounds like a good reason not to touch that this near release!
Labelling v20 and backport needed, so can find for v19.1 if appropriate

Karlson2k commented:
Thanks for @yol, I was able to remove discussable item.

@wsnipex, please check it. There is no more `libcdio++` in `libiso9660++` detection. :)

wsnipex commented:
I don't see how this resolves the issues from the other PR

Karlson2k commented:
@wsnipex Your position is counter-productive. Be concrete.

For a very simple fix I've wasted already ridiculous amount of time.
Instead of making Kodi better, I have to spend my time on a pointless tasks.

The second commit is not needed as it should be handled by host package dependencies.

wsnipex commented:
you keep ignoring everything I said - including code suggestions - from https://github.com/xbmc/xbmc/pull/19230
I'm not going to repeat it over and over

Karlson2k commented:
@wsnipex you want to fix other issues, not related. This PR is only to fix order of `libiso9660++` binaries. 
You are ignoring everything. The only clear position from your side is "I will not change my mind".

Write what is exactly wrong here, because I implemented already all reasonable corrections requested by you, including removal on `libcdio++` detection inside `iso9660++` module.

Karlson2k commented:
@wsnipex If you want to fix something else - OK, got ahead, make another PR. Don't squeeze unrelated changes here.

Karlson2k commented:
@wsnipex I've just rechecked especially for your. **ALL** your code suggestion are implemented.

wsnipex commented:
clearly not:
- libcdio++ headers are **required**
- cdio and cdio++ **must** be the same version -> fail if not

Karlson2k commented:
@wsnipex Clearly dependencies must be handled by package managers on OS. We should only detect presence.
We never check included headers on other libs. We should it be changed here?

Why not to check [all included headers in `cdio.h`](http://git.savannah.gnu.org/cgit/libcdio.git/tree/include/cdio/cdio.h?h=release-2.1.0#n35)? Let's add check for 
``` C
#include <cdio/version.h>
#include <cdio/types.h>
#include <cdio/sector.h>
```

This fix already improve order of `iso9660++` binaries and doesn't break anything.

wsnipex commented:
not sure what you are talking about:
we have a FindCdio module, but it currently does not check for the existence of libcdio++, which is a part of libcdio. -> this has nothing to do with distro packaging.

Karlson2k commented:
@wsnipex You confused with other modules.
This PR fixes only `libiso9660++`. 

FindCdio module finds `libcdio` and enable its functionality in Kodi. `libcdio++` is not used by Kodi and shouldn't be checked by FindCdio. Anyway, it is a separate topic and if you want to improve FindCdio modules, you can make another PR.

Karlson2k commented:
Jenkins build this please

wsnipex commented:
well, it doesn't fix it if libcdio++ existence is not verfied. iso9660pp requires libcdio++ for full functionality.
If this is not enforced, we end up with partially broken ISO playback, which is a nightmare from a support pov.

We're going around in circles and you keep insisting that kodi should be able to build without cdio++.
This is where I strongly disagree.

Karlson2k commented:
@wsnipex Please make some research before making any statements. 

`iso9660++` doesn't use `libcdio++` at all. Headers of `libcdio++` are included in `libiso9660++`, but that's it. No single function from `libcdio++` is called, no single object from `libcdio++` is used. Otherwise linker will fail, but it passed always just fine.

Kodi doesn't use `libcdio++` and doesn't need it. 
Topic is closed.

Karlson2k commented:
I have spent an absurdly large amount of time on a such simple and obvious fix.
I'm going to merge this PR as it brings fix (or at least improvement) and definitely makes no harm.
